src/libxively/control_topic/xi_control_message.c

```diff
@@ -0,0 +1,115 @@
+/* Copyright (c) 2003-2017, LogMeIn, Inc. All rights reserved.
+ *
+ * This is part of the Xively C Client library,
+ * it is licensed under the BSD 3-Clause license.
+ */
+
+#include <xi_control_message.h>
+#include <stddef.h>
+#include <xi_macros.h>
+#include <xi_helpers.h>
+
+#include <stdio.h>
+
+xi_control_message_t* xi_control_message_create_file_info( const char** filenames,
+                                                           const char** revisions,
+                                                           uint16_t count )
+{
+    if ( NULL == filenames || NULL == *filenames || 0 == count )
+    {
+        return NULL;
+    }
+
+    xi_state_t state = XI_STATE_OK;
+
+    XI_ALLOC( xi_control_message_t, file_info, state );
+
+    file_info->file_info.common.msgtype = XI_CONTROL_MESSAGE_DB_FILE_INFO;
+    file_info->file_info.common.msgver  = 1;
+
+    XI_ALLOC_BUFFER_AT( xi_control_message_file_desc_t, file_info->file_info.list,
+                        sizeof( xi_control_message_file_desc_t ) * count, state );
+
+    file_info->file_info.list_len = count;
+
+    uint16_t id_file = 0;
+    for ( ; id_file < count; ++id_file )
+    {
+        file_info->file_info.list[id_file].name = xi_str_dup( *filenames );
+        ++filenames;
+
+        if ( NULL != revisions )
+        {
+            file_info->file_info.list[id_file].revision = xi_str_dup( *revisions );
+            ++revisions;
+        }
+    }
+
+err_handling:
+
+    return file_info;
+}
+
+void xi_control_message_free( xi_control_message_t** control_message )
+{
+    if ( NULL == control_message || NULL == *control_message )
+    {
```

Comment:
Assert would crash the code, I wouldn;'t use it here. I think calling a free on a NULL pointer should be ok. Calling free multiple times on a pointer, I'd accept this.